### PR TITLE
use all_gather_into_tensor instead of all_gather

### PR DESCRIPTION
**Pull request by taozhiwei**
when using allgather, the output is a list, and in the implementation of torch, the list will be flattened and unflattened, which will result in additional allocation of GPU memory and D2D operations. But these all gather operations already have a flat GPU memory, using all_gather_into_tensor replaces all_gather will save GPU memory allocation and additional D2D operations.
additionally, batching all gatherers does not reduce the peak usage of GPU memory, so allgather_bucket_size has no effect.

### CODE REVIEW

**Review by GuanhuaWang — COMMENTED** [review]
Hi @taozhiwei , thx for making this PR, I think most parts are ok for me. please address those few comments, thx a ton!

File: deepspeed/ops/transformer/inference/moe_inference.py [hunk 1]
```diff
@@ -326,12 +326,12 @@ def forward(self,
...
-                ]
-                tensor_list[dist.get_rank(group=self.expert_mp_group)] = attention_output
-                dist.all_gather(tensor_list, attention_output, group=self.expert_mp_group)
-                attention_output = torch.cat(tensor_list).contiguous()
+                world_size = dist.get_world_size(group=self.expert_mp_group)
+                gather_buffer = torch.zeros(world_size * attention_output.numel(),
+                                            dtype=attention_output.dtype,
+                                            device=attention_output.device)
+                dist.all_gather_into_tensor(gather_buffer, attention_output, group=self.expert_mp_group)
+                attention_output = gather_buffer.view(-1, *attention_output.size()[1:])
```
[the remark this answers is not in the document]
Comment (taozhiwei) [reply deepspeed/ops/transformer/inference/moe_inference.py:334]: > sorry I might not be familiar with this `*attention_output.size()[1:]`, do you mean **1** `*attention_output.size()[1:]`? or you just want get values of dim 1+?

You can refer to this:[Unpacking Argument Lists](https://docs.python.org/3.8/tutorial/controlflow.html#unpacking-argument-lists)

File: deepspeed/runtime/utils.py [hunk 2]
```diff
@@ -937,38 +930,17 @@ def align_dense_tensors(tensor_list, alignment):
     return padded_tensor_list
 
 
-def all_gather_dp_groups(partitioned_param_groups, dp_process_group, start_alignment_factor, allgather_bucket_size):
```
[the remark this answers is not in the document]
Comment (taozhiwei) [reply deepspeed/runtime/utils.py:949]: > @tjruwase do we still want to maintain the original all-gather functionality with `nccl_start_alignment_factor` like this https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/runtime/utils.py#L955 ?

I think nccl_start_alignment_factor  is only needed when doing allgather in shards

[the remark this answers is not in the document]
Comment (tjruwase) [reply deepspeed/runtime/utils.py:949]: @GuanhuaWang, yes let's maintain for backwards-compatibility.

**Comment by tjruwase** [comment]
@taozhiwei, can you please check the overlap with this merged PR: https://github.com/microsoft/DeepSpeed/pull/4695

File: deepspeed/runtime/utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (GuanhuaWang) [reply deepspeed/runtime/utils.py:949]: > @GuanhuaWang, yes let's maintain for backwards-compatibility.

Hi @taozhiwei 

I am aligned here with @tjruwase , instead of delete older version, could you keep both your new version and older one? thanks

[the remark this answers is not in the document]
Comment (taozhiwei) [reply deepspeed/runtime/utils.py:949]: > > @GuanhuaWang, yes let's maintain for backwards-compatibility.
> 
> Hi @taozhiwei
> 
> I am aligned here with @tjruwase , instead of delete older version, could you keep both your new version and older one? thanks

I have already pushed the backward compatible code. could you please help me review it.thanks

**Comment by taozhiwei** [comment]
> @taozhiwei, can you please check the overlap with this merged PR: #4695

I have processed it and submitted the latest code.

File: deepspeed/runtime/utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (taozhiwei) [reply deepspeed/runtime/utils.py:949]: > > @GuanhuaWang, yes let's maintain for backwards-compatibility.
> 
> Hi @taozhiwei
> 
> I am aligned here with @tjruwase , instead of delete older version, could you keep both your new version and older one? thanks

@GuanhuaWang I have already pushed the backward compatible code. could you please help me review it.thanks

**Comment by GuanhuaWang** [comment]
just enabled our CI test.

**Review by GuanhuaWang — APPROVED** [review]
This PR has backward capability of all-gather with nccl start_alignment_factor, and provide new all-gather-into-tensor API. 

The new all_gather_into_tensor is applied on moe_inference, bf16_optimizer, and zero stage 1,2,3.

**Comment by taozhiwei** [comment]
> This PR has backward capability of all-gather with nccl start_alignment_factor, and provide new all-gather-into-tensor API.
> 
> The new all_gather_into_tensor is applied on moe_inference, bf16_optimizer, and zero stage 1,2,3.

Merging is blocked
The base branch restricts merging to authorized users.
@GuanhuaWang How should I solve this? thanks.
